### PR TITLE
Add a framework for typed InstIds.

### PR DESCRIPTION
**Pull request by jonmeow**
I'm switching `Temporary::storage_id` to a `TemporaryStorageInstId` as an example of this. Other than that, I'm not sure whether there are obvious low hanging fruit to change -- I figured starting to add the capability would be helpful towards use in future code.

Note, if you don't think this is more broadly helpful, I can abort this, but I think this is something you'd brought up so just trying  to provide it while feeling a little blocked elsewhere.

### CODE REVIEW

File: toolchain/sem_ir/id_kind.h [hunk 1]
```diff
@@ -115,16 +116,19 @@ class TypeEnum {
...
     // From sem_ir/id.h.
     InstId, AbsoluteInstId, MatchingInstId, ConstantId, EntityNameId,
     CompileTimeBindIndex, RuntimeParamIndex, FunctionId, ClassId, InterfaceId,
     ImplId, GenericId, SpecificId, ImportIRId, ImportIRInstId, LocId, BoolValue,
     IntKind, NameId, NameScopeId, InstBlockId, TypeId, TypeBlockId,
-    ElementIndex, LibraryNameId, FloatKind>;
+    ElementIndex, LibraryNameId, FloatKind,
+// Subtyped InstId variants.
+#define CARBON_SEM_IR_INST_KIND(Name) Name##InstId,
+#include "toolchain/sem_ir/inst_kind.def"
```
Comment (zygoloid) [inline toolchain/sem_ir/id_kind.h:129]: I'm concerned by this part: right now we have a few places where we're uniformly processing `InstId` operands of instructions by looking at the `IdKind`, and after this change those places will no longer visit the `storage_id` field on `Temporary` ([example](https://github.com/carbon-language/carbon-lang/blob/85f6bf32b5a42fcd5d0ac5faa4e4780a68787710/toolchain/check/subst.cpp#L113), [example](https://github.com/carbon-language/carbon-lang/blob/85f6bf32b5a42fcd5d0ac5faa4e4780a68787710/toolchain/check/deduce.cpp#L105)). Maybe we could map all the typed `InstId` kinds to `IdKind::For<InstId>`?

**Review by zygoloid — COMMENTED** [review]
I think this could be useful. Do you know how much this affects compile times? (It looks like it shouldn't be much to me.)

**Comment by jonmeow** [comment]
> I think this could be useful. Do you know how much this affects compile times? (It looks like it shouldn't be much to me.)

Right now it doesn't affect much. Maybe I'm a little biased due to typed node IDs and extract.cpp performance, but I think the biggest risk for compile performance is if it sees broad use in typed insts. Whereas right now that's not an issue, the number of kinds they deal with right now is small. If these start seeing broad use in signatures, I would expect that the things dealing with typed inst reflection will slow down (due to less reuse), and overall we could see a significant slowdown in compile times. However, it's possibly premature to make assumptions in that direction.

**Comment by jonmeow** [comment]
Per discussion, closing this PR. We went through current uses of `As` and `GetAs`, and it didn't look like there were enough spots where a `TypedInstId` could simplify things. We may look into specialized `StructTypeField` storage,  but that wouldn't have been helped here unless we had typed `InstBlockId` fetching too.

File: toolchain/sem_ir/id_kind.h (hunk 1 above)
Comment (jonmeow) [reply toolchain/sem_ir/id_kind.h:129]: PS: I was concerned that collapsing to `IdKind::For<InstId>` would affect the ability of `NodeStack` (and possibly others) to propagate a `TypedInstId`. This is what led the discussion below, where there's maybe not actually enough value, closing the PR.

**Pull request closed without merging**